HISTORY.md

```diff
@@ -2,6 +2,7 @@
 ## Unreleased
 ### New Features
 * Add a new option OptimisticTransactionDBOptions::shared_lock_buckets that enables sharing mutexes for validating transactions between DB instances, for better balancing memory efficiency and validation contention across DB instances. Different column families and DBs also now use different hash seeds in this validation, so that the same set of key names will not contend across DBs or column families.
+* Start logging non-zero user-defined timestamp sizes in WAL to signal user key format in subsequent records and use it during recovery.
```

Comment:
One thing to note is that this change breaks recovery from WAL files containing timestamps written by earlier versions that don't use timestamp size markers. The UDTS feature is still marked experimental so this should be fine but might still want to put a warning here for existing OSS users of the feature. The workaround would be to ensure there are no WAL files to recover (i.e. by flushing before close) before upgrading.

Reply:
Thank you for bringing this up! It's great to call out these compatibility warnings. I liked your description of the issue, so I kinda of copied from your comment.

Reply:
I tried using it end-to-end to see if I'd run into any mention of it being experimental, but didn't. Those APIs were:

- `BytewiseComparatorWithU64Ts`
- `DB::Put()` (an overload taking `ts`)
- `ReadOptions::timestamp`

But the feature has existed for years and is covered by correctness testing so probably doesn't need to be marked experimental

Reply:
You are right the feature probably shouldn't be treated as experimental anymore since it's already functional and tested. Let me update that. I will be more careful for introducing incompatible changes.

Reply:
> I tried using it end-to-end to see if I'd run into any mention of it being experimental, but didn't.

Not sure about the API docs but it's still marked as experimental on the wiki: https://github.com/facebook/rocksdb/wiki/User-defined-Timestamp-%28Experimental%29

> But the feature has existed for years and is covered by correctness testing so probably doesn't need to be marked experimental

IIRC the reason we've kept it experimental is to allow for the occasional breaking change (like this one, or when we completely changed the write APIs a while back). But yeah, I agree that we can probably remove the label now that the format compatibility issue got solved by this PR.